dh.c

```diff
@@ -152,11 +155,52 @@ choose_dh(int min, int wantbits, int max)
 	int best, bestcount, which, linenum;
 	struct dhgroup dhg;
 
+#ifndef WINDOWS
 	if ((f = fopen(_PATH_DH_MODULI, "r")) == NULL) {
 		logit("WARNING: could not open %s (%s), using fixed modulus",
 		    _PATH_DH_MODULI, strerror(errno));
 		return (dh_new_group_fallback(max));
 	}
+#else
+	/* First check the moduli file in the %programdata%\ssh\ directory.
+	 * If not then search for the moduli file in the current executable directory. This file will be updated in new OpenSSH releases.
+	 */
+	if ((f = fopen(_PATH_DH_MODULI, "r")) == NULL) {
+		debug3("Could not open %s (%s)",
+			_PATH_DH_MODULI, strerror(errno));
+
+		int isFallback = 1;
+		extern char* __progdir;
+		if (__progdir) {
+			char moduli_path[PATH_MAX] = { 0 };
+			_snprintf_s(moduli_path, PATH_MAX, _TRUNCATE, "%s\\moduli", __progdir);
+
+			if ((f = fopen(moduli_path, "r")) == NULL) {
+				debug3("Could not open %s (%s)", moduli_path, strerror(errno));
+			} else {
+				if (check_secure_file_permission(moduli_path, NULL, 1) != 0) {
```

Comment:
if t moduli file is deployed by manifest. the ACL of the file will look like below. Please make sure this function invocation work.
![image](https://user-images.githubusercontent.com/12284203/76877156-3bd8bc00-6830-11ea-96fa-344f89c5661f.png)

Reply:
Validated. The code works fine with these ACLs.